game-core/src/main/java/games/strategy/triplea/delegate/dice/roller/RollBattleDice.java

```diff
@@ -0,0 +1,41 @@
+package games.strategy.triplea.delegate.dice.roller;
+
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.dice.calculator.LowLuckDice;
+import games.strategy.triplea.delegate.dice.calculator.RolledDice;
+import games.strategy.triplea.delegate.power.calculator.CombatValue;
+import games.strategy.triplea.delegate.power.calculator.PowerStrengthAndRolls;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.Collection;
+import lombok.experimental.UtilityClass;
+
+/** Used to roll dice for attackers and defenders in battles. */
+@UtilityClass
+public class RollBattleDice {
+
+  public static DiceRoll rollDice(
+      final Collection<Unit> units,
+      final GamePlayer player,
+      final IDelegateBridge bridge,
+      final String annotation,
+      final CombatValue combatValueCalculator) {
+
+    final DiceRoll diceRoll;
+    final PowerStrengthAndRolls unitPowerAndRollsMap =
+        PowerStrengthAndRolls.build(units, combatValueCalculator);
+    if (Properties.getLowLuck(bridge.getData().getProperties())) {
+      diceRoll = LowLuckDice.calculate(unitPowerAndRollsMap, player, bridge::getRandom, annotation);
+    } else {
+      diceRoll = RolledDice.calculate(unitPowerAndRollsMap, player, bridge::getRandom, annotation);
+    }
+
+    bridge
```

Comment:
I don't know if the bridge call belongs here. I would personally opt to duplicate this bridge call wherever we use the dice roller. The reasoning is largely just SRP, we are doing a lot of things here and the bridge call is a side effect.

Reply:
Looking at what remains, I'm not really convinced this is a useful class. Can this be combined with the RollNDice class? It seems like this should be a stateful 'diceRoller' class and then it handles whether it rolls LL or normal dice as an implementation detail.

Reply:
I am concerned this is going to get confusing, when does one use `RollBattleDice` and when do you use `RollNDice`? I think combining and making these two classes potentially stateful would be a good way to go. I would seriously lean to remove the bridge code as that is side-effect of reporting the result and is not intrinsic to rolling dice.

Reply:
The bridge call is unfortunately intrinsic to dice.  When you save a game and then reload it, it will calculate the dice probability of the saved game using these history items.  If they are missing or not written correctly, then the dice probability of the saved game will be invalid.

I don't agree with the design but I don't see how to easily change that.  And since every call of these methods needs to also write the history, I think it is best to leave the call in here so that it isn't forgotten.  I'll add a comment to the code so that it explains why this history is required.

Reply:
These three classes (`RollAaDice`, `RollNDice`, and `RollBattleDice`) are basically strategy objects that define what type of dice you are getting.  I could see creating a builder class that you pass in an enum but I'm not sure how that is any better than just calling the static method itself.

Reply:
I have been thinking of changing the annotation from a `String` to a `DiceAnnotation` (or similar).  Then the parameter to these methods would require a `DiceAnnotation` instead of a `String` and it would ensure that the annotation is the correct format (since the history string is parsed to build the probability history).  I could make `DiceAnnotation` accept a bridge so that it also writes the history and then change these classes to call a method on the `DiceAnnotation` object that will write the history.  But that would be another PR.

Reply:
My 2 cents, if anyone is using the dice roll code then they'll be copy/pasting from an existing usage and the bridge will be there. I will admit there is potential for error. 

On the other hand, if you want to get 'just a dice roll', it's tightly bound to the bridge call and it's a side-effect that someone would not want. I could imagine a duct-tape-programmer putting in a boolean parameter to avoid this side-effect.

With that said, leaving it for now is okay. I like your ideas for further improvements.

[resolve]